branches/sam/hfst3-pmatchdev/man/hfst-xfst-compiler.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.37.1.
.TH HFST-XFST-COMPILER "1" "November 2010" "HFST" "User Commands"
.SH NAME
hfst-xfst-compiler \- manual page for hfst-lexc-compiler 0.1
.SH SYNOPSIS
.B hfst-xfst-compiler
[\fIOPTIONS\fR...] [\fIINFILE1\fR...]]
.SH DESCRIPTION
Compile xfst files into automata
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.SS "Input/Output options:"
.TP
\fB\-f\fR, \fB\-\-format\fR=\fIFORMAT\fR
compile into FORMAT transducer
.PP
If INFILE is omitted or \-, standard stream will be used
FORMAT must be one of supported transducer formats, such as openfst\-tropical, sfst, foma, etc.
.SH EXAMPLES
.TP
hfst\-xfst\-compiler cat.xfst
Compile single\-file script
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://sourceforge.net/tracker/?atid=1061990&group_id=224521&func=browse>
.SH "SEE ALSO"
hfst\-xfst\-compiler home page:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstXfstCompiler>
.br
General help using HFST software:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstHome>
.SH COPYRIGHT
Copyright \(co 2010 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.3.9
---------------------------

* Bugs fixed and new features added to hfst-proc

* Xfst rules tentatively supported in hfst-regexp2fst

* hfst-fst2strings does not produce duplicate paths any more 
  when using foma as backend

* hfst-lexc implementation format defaults to openfst-tropical
  instead of foma

Noteworthy changes in 3.3.8
---------------------------

* Fixes 3.3.7

Noteworthy changes in 3.3.7
---------------------------

* Compatible with OpenFst version 1.3.1

* Identity symbols supported in hfst-regexp2fst

Noteworthy changes in 3.3.6
---------------------------

* Include directives of foma and SFST header files changed so
  that standard include path is used instead of source path

* Compilation warnings and some memory leaks fixed

Noteworthy changes in 3.3.5
---------------------------

* hfst-fst2txt.cc also writes to dot/graphviz format

* Epenthesis added for Xerox rules parser

* hfst-compose-intersect gives warnings on missing symbols

Noteworthy changes in 3.3.4
---------------------------

* Substitution functions optimized and new substitution functions added that
  perform multiple substitutions at the same time.

* The tool hfst-substitute now implements the option --in-order.

* Updated to OpenFst version 1.2.10.

* Conversions between transducer formats are more efficient.

Noteworthy changes in 3.3.3
---------------------------

* HfstTransitionGraph and optimized lookup transducer code optimized

* Foma transducer name is not copied to the net but is left empty

* Small fixes to twolc

Noteworthy changes in 3.3.2
---------------------------

* Forward compatibility with OpenFst 1.2.8 is achieved.

* PKG_ macros removed from end user builds since they do not work nicely on
  all platforms.

Noteworthy changes in 3.3.1
---------------------------

* Handling of flag diacritics improved.

* More examples in fsmbook-examples included in tests.

* Extra files created during tests as well as generated autotools files 
  are now ignored by svn.

Noteworthy changes in 3.3.0
---------------------------

* HFST updated to work with the newest versions of backend libraries

* a new replace down function by Karttunen added

* HfstTransitionGraph and conversions between formats made less error-prone

* protected/public changes in API

* SFST and foma libraries are disabled from default build

Noteworthy changes in 3.2.0
---------------------------

* Numerous improvements to hfst-proc, including better uppercasing via glib,
  better compound handling, new analysis mode for raw analysis without case
  changes or special symbols etc.

* old slow lexc (not built by default)

* numerous scripting fixes for mac os x compatibility with foma

* constness changes in API

Noteworthy changes in 3.1.1
---------------------------

* Numerous bugfixes, in particular regarding a broken assumption in
  compose-intersect

Noteworthy changes in 3.1.0
---------------------------

* api for lookup and lookdown changed, previously memory didn't get initialized
  anywhere, causing a serious bug

Noteworthy changes in 3.0.3
---------------------------

* numerous internal improvements, eg. distcheck is now functional

* many bugfixes, especially wrt. INDENTITY_PAIR & OTHER SYMBOL, composition,
  final weights in ol-converted transducers

Noteworthy changes in 3.0.2
---------------------------

* some fixes to OSX compatibility

* no more license conflict problems wrt foma

* better memory use in compose-intersect

* new implementation of internal transducer

Noteworthy changes in 3.0.1
---------------------------

* this is a bugfix release

* conversion to hfst-optimized-lookup format was fundamentally broken in the
  original release; this should provide a fix and considerable speedup

* compose-intersect memory use is much improved

Noteworthy changes in 3.0
-------------------------

* all new API in object oriented form

* brand new foma backend for unweighted transducer handling

* the FST libraries used for automata handling are now dynamically linked to
  HFST

    * this requires end users to install needed backends themselves

* hfst-lexc and hfst-xfst based on foma's compiler

  * xfst is now also fully supported

  * the obsolete command line clients for HFST-based compilation of lexc and
    xfst scripts are hidden in hfst-lexc2fst and hfst-xfst2fst

* preliminary support for unknown character classes

* optimized-lookup backend supported in library

  * dropped special tools for optimised handling since optimised lookup is
    included in library: hfst-optimised-lookup -> hfst-lookup; 
    hfst-lookup-optimize -> hfst-fst2fst

* hfst-proc tool to emulate lt-proc, cg-proc

* Slightly improved string and text formats to support weights uniformly and
  parsing tabs and spaces as automata arcs: see `wiki page for text and string
  formats
  <https://kitwiki.csc.fi/twiki/bin/view/KitWiki/HfstStringAndTextFormats>`

* better error handling using exceptions

* The argument handling order of binary tools changed: stdin is first if not
  specified

* The automata are now packed in HFST automaton container format (akin to AVI
  for videos); this supports arbitrary metadata and underlying FST formats

* the symbol handling is more implicit; no more symbol table arguments for tools
  or hfst-symbols command

* converted to new autotools: silent rules and color tests

* Hfst's internal automata format for fallback on some unsupported operations
  and for conversions

Noteworthy changes in 2.4.1
---------------------------

* Fixed inclusion path in pkg-config file

* Fixed extraneous epsilon path in hfst-lexc transducers

Noteworthy changes in release 2.4
---------------------------------


hfst3/tools/src/hfst-xfst-compiler.cc
//! @file hfst-xfst-compiler.cc
//!
//! @brief Xfst compilation command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif


#include <iostream>
#include <fstream>

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>

#include "HfstTransducer.h"
#include "parsers/XfstCompiler.h"

using hfst::HfstTransducer;
using hfst::HfstInputStream;
using hfst::HfstOutputStream;
using hfst::ImplementationType;

using hfst::xfst::XfstCompiler;

#include "hfst-commandline.h"
#include "hfst-program-options.h"


#include "inc/globals-common.h"
static char** xfstfilenames = 0;
static FILE** xfstfiles = 0;
static unsigned int xfstcount = 0;
static bool is_input_stdin = true;
static ImplementationType format = hfst::UNSPECIFIED_TYPE;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE1...]]\n"
             "Compile xfst files into transducer\n"
        "\n", program_name );
        print_common_program_options(message_out);
        fprintf(message_out, "Input/Output options:\n"
               "  -f, --format=FORMAT     compile into FORMAT transducer\n"
               "  -o, --output=OUTFILE    write result into OUTFILE\n");
        fprintf(message_out, "\n");
        fprintf(message_out,
                "If INFILE or OUTFILE are omitted or -, standard streams will "
                "be used\n"
		"The possible values for FORMAT are { sfst, openfst-tropical, "
		"openfst-log,\n"
		"foma, optimized-lookup-unweighted, optimized-lookup-weighted }.\n"
        fprintf(message_out,
            "\n"
            "Examples:\n"
            "  %s -o cat.hfst cat.xfst               Compile single-file "
            "lexicon\n"
            "  %s -o L.hfst Root.xfst 2.xfst 3.xfst  Compile multi-file "
            "lexicon\n"
            "\n",
            program_name, program_name );
        print_report_bugs();
        fprintf(message_out, "\n");
        print_more_info();
}

int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
          HFST_GETOPT_COMMON_LONG,
          {"format", required_argument, 0, 'f'},
          {"output", required_argument, 0, 'o'},
          {0,0,0,0}
        };
        int option_index = 0;
        char c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             "f:o:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }
        switch (c)
        {
#include "inc/getopt-cases-common.h"
        case 'f':
          format = hfst_parse_format_name(optarg);
          break;
#include "inc/getopt-cases-error.h"
        }
    }

#include "inc/check-params-common.h"
    if (format == hfst::UNSPECIFIED_TYPE)
      {
        verbose_printf("Defaulting to OpenFst tropical type\n");
        format = hfst::TROPICAL_OPENFST_TYPE;
      }
    if (argc - optind > 0)
      {
        xfstfilenames = static_cast<char**>(malloc(sizeof(char*)*(argc-optind)));
        xfstfiles = static_cast<FILE**>(malloc(sizeof(char*)*(argc-optind)));
        while (optind < argc)
          {
            xfstfilenames[xfstcount] = hfst_strdup(argv[optind]);
            xfstfiles[xfstcount] = hfst_fopen(argv[optind], "r");
            xfstcount++;
            optind++;
          }
        is_input_stdin = false;
      }
    else
      {
        xfstfilenames = static_cast<char**>(malloc(sizeof(char*)));
        xfstfiles = static_cast<FILE**>(malloc(sizeof(FILE*)));
        xfstfilenames[0] = hfst_strdup("<stdin>");
        xfstfiles[0] = stdin;
        is_input_stdin = true;
        xfstcount++;
      }
    return EXIT_CONTINUE;
}

int
xfst_streams(XfstCompiler& xfst, HfstOutputStream& outstream)
{
  //outstream.open();
    for (unsigned int i = 0; i < xfstcount; i++)
      {
        verbose_printf("Parsing xfst file %s\n", xfstfilenames[i]);
        if (xfstfiles[i] == stdin)
          {
            verbose_printf("hfst[0]: ");
            xfst.parse(stdin);
          }
        else
          {
            xfst.parse(xfstfilenames[i]);
          }
      }
    outstream.close();
    return EXIT_SUCCESS;
}


int main( int argc, char **argv ) {
    hfst_set_program_name(argv[0], "0.1", "HfstXfstCompiler");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    for (unsigned int i = 0; i < xfstcount; i++)
      {
        if (xfstfiles[i] != stdin)
          {
            fclose(xfstfiles[i]);
          }
      }
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from ");
    for (unsigned int i = 0; i < xfstcount; i++)
      {
        verbose_printf("%s, ", xfstfilenames[i]);
      }
    verbose_printf("writing to %s\n", outfilename);
    // here starts the buffer handling part
    HfstOutputStream* outstream = (outfile != stdout) ?
        new HfstOutputStream(outfilename, format) :
        new HfstOutputStream(format);
    XfstCompiler xfst(format);
    if (silent)
      {
        xfst.setVerbosity(false);
      }
    else
      {
        xfst.setVerbosity(verbose);
      }
    retval = xfst_streams(xfst, *outstream);
    delete outstream;
    free(xfstfilenames);
    free(outfilename);
    return retval;
}

